Update solvers to handle binomial.

sbt4104 opened the pull request:
#### References to other Issues or PRs
Fixes #10993

#### Brief description of what is fixed or changed
Added conditions to expand binomial in solvers
Update expand function for binomial
#### Other comments 
Please suggest changes.

### Release Notes
<!-- BEGIN RELEASE NOTES -->
* solvers
    - Fix solvers to handle binomials
* functions
    - Update expand function for binomials.
<!-- END RELEASE NOTES -->

sbt4104 committed:
Update solvers to handle binomial.

sbt4104 committed:
Update solvers to handle binomials.

oscarbenjamin commented:
I'm not sure where this belongs but certainly not as the opening lines in the solve function. Unfortunately solve is a bit of a mess so there isn't necessarily a correct place to put something like this.

sbt4104 committed:
Place condition in preprocess equation.

sbt4104 commented:
I think putting those conditions in "preprocess equations" is a more right.

sbt4104 committed:
Place condition in preprocess equation(remove whitespaces).

sbt4104 commented:
Please review and suggest the changes.

oscarbenjamin commented:
Consider that you might have an expression involving a binomial like
```
In [4]: expr = 1 + binomial(x, 3)                                                                                                                             

In [5]: isinstance(expr, binomial)                                                                                                                            
Out[5]: False

In [6]: expr.has(binomial)                                                                                                                                    
Out[6]: True
```
We probably don't want to expand all functions so it would be better to use expr.replace to pick out the binomials and expand them.

This code is still not in the right place: solve hasn't even converted Eqs to Expr at that point. Look through solve to find where similar transformations are happening.

sbt4104 committed:
changed solver to expand only binomials from expression.

sbt4104 committed:
changed solver to expand only binomials from expression.

sbt4104 commented:
I have made the changes so that only binomial terms are expanded and added the code after Eqs are checked and converted. Please check if it is correctly done.

sbt4104 committed:
whitespaces removed.

sbt4104 commented:
can someone explain to me why my test on travis-ci is failing.
`$ bin/test_travis.sh`
` +[[ '' == \t\r\u\e ]] `
 `+[[ '' == \t\r\u\e ]] `
 `+[[ '' == \t\r\u\e ]] `
 `+[[ '' == \t\r`

sbt4104 committed:
changed solver to expand only binomials from expression.

sbt4104 commented:
I pushed the same changes again and it passed.
Please review the changes.

oscarbenjamin commented on sympy/solvers/solvers.py:
```diff
@@ -999,6 +999,9 @@ def _sympified_list(w):
             f[i] = fi
 
+        if fi.has(binomial):
+            f[i] = fi.replace(lambda exp: binomial, lambda exp: expand_func(exp))
```
The filter function here will return true for any expression:
```
In [1]: f = lambda exp: binomial                                                                                                               

In [2]: f(sin(x))                                                                                                                              
Out[2]: binomial

In [3]: bool(f(sin(x)))                                                                                                                        
Out[3]: True
```

sbt4104 replied:
I have made the change for it to correctly identify the binomials.

sbt4104 committed:
changed filter to identify only binomial.

oscarbenjamin commented:
I think translating binomials should still come later after the loop that it is in now.

This will need to have a test added for equations with binomial coefficients that can be solved.

sbt4104 commented:
I will make the changes.

sbt4104 committed:
Add test.

sbt4104 commented:
I have added the test.

sbt4104 commented:
I have added the binomial expansion after the loop, please suggest changes.

oscarbenjamin commented on sympy/solvers/solvers.py:
```diff
@@ -1027,6 +1027,11 @@ def _sympified_list(w):
+    #expand binomial terms
+    for i, fi in enumerate(f):
+        if fi.has(binomial):
+            f[i] = fi.replace(lambda exp: binomial, lambda exp: expand_func(exp))
```
The filter function here will still match non-binomial expressions.

sbt4104 replied:
I will correct it right away.

oscarbenjamin replied:
How about:
```python
expand_binomial = lambda f: f.replace(binomial, lambda n, k: expand_func(binomial(n, k)))
f = [expand_binomial(fi) for fi in f]
```

sbt4104 replied:
This one looks more efficient, I will make the changes.

sbt4104 replied:
I earlier thought that binomial expansion should be done in the previous loop as hyperbolic functions are checked and rewritten inside it.

sbt4104 replied:
please review the changes.

sbt4104 committed:
Update solver to handle binomials.

sbt4104 committed:
Update solvers.

oscarbenjamin commented:
On master we have:
```julia
In [1]: solve(x-binomial(y, 10))                                                                                                                              
Out[1]: 
⎡⎧   ⎛y ⎞⎫⎤
⎢⎨x: ⎜  ⎟⎬⎥
⎣⎩   ⎝10⎠⎭⎦
```
This PR gives
```julia
In [1]: solve(x-binomial(y, 10))                                                                                                                              
Out[1]: 
⎡⎧     ⎛ 9       8        7         6          5           4           3            2                     ⎞⎫⎤
⎢⎪   y⋅⎝y  - 45⋅y  + 870⋅y  - 9450⋅y  + 63273⋅y  - 269325⋅y  + 723680⋅y  - 1172700⋅y  + 1026576⋅y - 362880⎠⎪⎥
⎢⎨x: ──────────────────────────────────────────────────────────────────────────────────────────────────────⎬⎥
⎢⎪                                                  3628800                                                ⎪⎥
⎣⎩                                                                                                         ⎭⎦
```
I think the behaviour on master is better so I guess the replace call needs to be modified so that it only expands binomials that involve the unknowns that we are trying to solve for.

sbt4104 committed:
expand binomials only if it has the unknown symbol.

sbt4104 commented:
I have made the changes, please check.

sbt4104 commented:
It is failing if in binomial(n,k) n is a polynomial in x, example
`y = solve(1-binomial(x+1, 3),x)`
I think adding this will remove this error
`f = f.replace(binomial, lambda n, k: expand_func(binomial(n, k)) if symbol in n.free_symbols and len(n.free_symbols) == 1 else binomial(n,k))`

sbt4104 commented:
@oscarbenjamin should I make this change?

sbt4104 committed:
check if n is a polynomial having only the unknown symbol.

sbt4104 committed:
add tests

oscarbenjamin commented:
> It is failing if in binomial(n,k) n is a polynomial in x, example
> `y = solve(1-binomial(x+1, 3),x)`

In what way was it failing?

sbt4104 commented:
I had written it like this in the previous commit
`f = f.replace(binomial, lambda n, k: expand_func(binomial(n, k)) if n==symbol else binomial(n, k))`
so for `y = solve(1-binomial(x+1, 3),x)`, it was comparing x+1 with x and returned 'NotImplementedError: 
No algorithms are implemented to solve equation binomial(x + 3, 3)'

 though it worked perfectly for  `y = solve(1-binomial(x+1, 3), x+1)`


so i changed it to solve cases like this 
`f = f.replace(binomial, lambda n, k: expand_func(binomial(n, k)) if symbol in n.free_symbols and len(n.free_symbols) == 1 else binomial(n,k))`

oscarbenjamin commented:
This is probably best:
```
In [5]: syms = [x]                                                                                                                                            

In [6]: binomial(x, 2).replace(lambda e: isinstance(e, binomial) and e.has(*syms), lambda e: expand_func(e))                                                  
Out[6]: 
x⋅(x - 1)
─────────
    2  
```

oscarbenjamin commented:
It looks like `expand_func` for binomial could be improved:
```
In [18]: expand_func(binomial(x, 2))                                                                                                                          
Out[18]: 
x⋅(x - 1)
─────────
    2    

In [19]: expand_func(binomial(x+2, x))                                                                                                                        
Out[19]: 
⎛x + 2⎞
⎜     ⎟
⎝  x  ⎠

In [20]: gammasimp(_.rewrite(factorial))                                                                                                                      
Out[20]: 
(x + 1)⋅(x + 2)
───────────────
       2   
```

sbt4104 commented:
> This is probably best:
> 
> ```
> In [5]: syms = [x]                                                                                                                                            
> 
> In [6]: binomial(x, 2).replace(lambda e: isinstance(e, binomial) and e.has(*syms), lambda e: expand_func(e))                                                  
> Out[6]: 
> x⋅(x - 1)
> ─────────
>     2  
> ```

Should we expand the binomial if we have a multivariable equation, example
 `binomial(x + y, 2)`

oscarbenjamin commented:
> Should we expand the binomial if we have a multivariable equation

Yes if any of the symbols of interest is in the binomial. Otherwise we won't be able to solve the equations.

sbt4104 commented:
ok, I will make that change.

sbt4104 committed:
expand binomial having unknown symbol.

sbt4104 commented:
I have made the changes, please review.

sbt4104 commented:
> It looks like `expand_func` for binomial could be improved:
> 
> ```
> In [18]: expand_func(binomial(x, 2))                                                                                                                          
> Out[18]: 
> x⋅(x - 1)
> ─────────
>     2    
> 
> In [19]: expand_func(binomial(x+2, x))                                                                                                                        
> Out[19]: 
> ⎛x + 2⎞
> ⎜     ⎟
> ⎝  x  ⎠
> 
> In [20]: gammasimp(_.rewrite(factorial))                                                                                                                      
> Out[20]: 
> (x + 1)⋅(x + 2)
> ───────────────
>        2   
> ```

expand_func expands `binomial(x, x-1)` properly but fails for `binomial(x+1, x)`

I think changing 
`if k.is_Add and n in k.args : k = n - k` 
to
`if (k.is_Add and n in k.args) or (n.is_Add and k in n.args) : k = n - k`
will solve this problem

oscarbenjamin commented:
Or maybe just `if (k - n).is_Integer`

sbt4104 commented:
yes, this one will also expand cases like
`solve_func(binomial(x + 1, x - 1))`

sbt4104 commented:
should I send a PR for this along with the issue?

oscarbenjamin commented:
You can make those changes in this PR as well

sbt4104 committed:
Update expand for binomials.

sbt4104 commented:
I have made the changes, please review.

oscarbenjamin commented on sympy/functions/combinatorial/factorials.py:
```diff
@@ -1012,7 +1012,7 @@ def _eval_expand_func(self, **hints):
         k = self.args[1]
-        if k.is_Add and n in k.args:
+        if (n-k).is_Integer:
             k = n - k
```
There should be a test for this.

oscarbenjamin commented on sympy/solvers/solvers.py:
```diff
@@ -1460,6 +1460,9 @@ def _solve(f, *symbols, **flags):
     symbol = symbols[0]
 
+    #expand binomials only if it has the unknown symbol
+    f = f.replace(lambda e: isinstance(e, binomial) and e.has(symbol), lambda e: expand_func(e))
```
This line is too long

oscarbenjamin commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -2137,6 +2137,13 @@ def test_issue_17949():
+    assert solve(Eq(binomial(x,2),0)) == [0, 1]
+    assert solve(a+binomial(x, 3),a) == [-binomial(x, 3)]
+    assert solve(x-binomial(a, 3) + binomial(y,2) + sin(a),x) == [-sin(a) + binomial(a, 3) - binomial(y, 2)]
+    assert solve((x+1)-binomial(x+1, 3),x) == [-2, -1, 3]
```
Include a space after a comma

sbt4104 committed:
Proper spaces after comma.

sbt4104 committed:
add test

sbt4104 committed:
add test.

sbt4104 committed:
add tests.

oscarbenjamin commented:
This looks good. I'll merge it once tests pass.

For future PRs can you try to use better commit messages? The messages here started out good but then got less good as time went on. You should imagine that the message will be looked at in future by someone who does not know the full context of the PR. Something like "add test" does not convey much about the change that is being made to the code.

sbt4104 commented:
I will keep it in mind, this will not happen again.

oscarbenjamin commented:
Merged. Thanks for this!

sbt4104 commented:
Thank you, l learned a lot from this.
